emote/callbacks.py

```diff
@@ -424,25 +431,43 @@ def __init__(
         path: str,
         checkpoint_index: int,
         reset_training_steps: bool = False,
-        optimizers: Optional[List[optim.Optimizer]] = None,
-        networks: Optional[List[nn.Module]] = None,
+        only_load_nets: bool = False,
     ):
         super().__init__()
-        self._cbs = callbacks
         self._path = path
         self._checkpoint_index = checkpoint_index
         self._reset_training_steps = reset_training_steps
-        self._opts: List[optim.Optimizer] = optimizers if optimizers else []
-        self._nets: List[nn.Module] = networks if networks else []
+        self._only_load_nets = only_load_nets
+
+        self._cbs = []
+        names = []
+        for cb in callbacks:
+            if hasattr(cb, "name"):
+                self._cbs.append(cb)
+                names.append(cb.name)
+            else:
+                warnings.warn(
+                    f"CheckpointLoader ignored {cb} because of not "
+                    f"having the 'name' field.",
+                    UserWarning,
+                )
+
+        if len(names) != len(set(names)):
+            raise ValueError(
+                "CheckpointLoader is given a list of callbacks with at least"
+                "two callbacks with identical names"
+            )
 
     def begin_training(self):
         state_dict: dict = torch.load(f"{self._path}.{self._checkpoint_index}.tar")
-        for cb, state in zip(self._cbs, state_dict["callback_state_dicts"]):
-            cb.load_state_dict(state)
-        for net, state in zip(self._nets, state_dict["network_state_dicts"]):
-            net.load_state_dict(state)
-        for opt, state in zip(self._opts, state_dict["optim_state_dicts"]):
-            opt.load_state_dict(state)
+
+        for cb in self._cbs:
+            state = state_dict["callback_state_dicts"][cb.name]
+            if self._only_load_nets and hasattr(cb, "network"):
+                cb.network.load_state_dict(state.pop("network_state_dict"))
+            else:
+                cb.load_state_dict(state)
```

Comment:
This else will always load for things that don't have a network. I don't think that's intended. I'd consider passing the flags into the load_state_dict instead. Maybe making it a dataclass with a few fields? The ones we've traditionally used for flagging is the following:

```
@dataclass
class LoadConfiguration:
    """Load the network weights"""

    load_weights: bool = True

    """Load the hparams file containing our steps and other non-ML state"""
    load_hparams: bool = True

    """Load the optimizer state from the checkpoint"""
    load_optimizer_state: bool = True
```

We don't necessarily we have to use the same fields or class, just that these are the three toggles we've traditionally had.

Reply:
I think my concern here is that if I somehow call it _network or have two networks it's going to lead to weird bugs. Having this knowledge in the callback makes more sense.